op-node/l1/batching.go

```diff
@@ -2,134 +2,181 @@ package l1
+	ibc.resetLock.RLock()
+	defer ibc.resetLock.RUnlock()
+
+	// collect a batch from the requests channel
+	batch := make([]rpc.BatchElem, 0, ibc.batchSize)
+	// wait for first element
+	select {
+	case reqElem, ok := <-ibc.scheduled:
+		if !ok { // no more requests to do
+			return io.EOF
 		}
-		return i
+		batch = append(batch, reqElem)
+	case <-ctx.Done():
+		return ctx.Err()
 	}
 
-	maxCount := expectedBatches * maxRetry
-
-	awaited := len(requests)
-
-	// start initial round of parallel requests
-	count := parallelRequests()
-
-	// We slow down additional batch requests to not spam the server.
-	retryTicker := time.NewTicker(time.Millisecond * 20)
-	defer retryTicker.Stop()
-
-	// The main requests slice is only ever mutated by the go-routine running this loop.
-	// Slices of this are sent to worker go-routines, and never overwritten with different requests.
+	// collect more elements, if there are any.
 	for {
-		// check if we've all results back successfully
-		if awaited <= 0 {
-			return nil
-		}
-		if count > maxCount {
-			return TooManyRetries
+		if len(batch) >= ibc.batchSize {
+			break
 		}
 		select {
-		case <-retryTicker.C:
-			count += parallelRequests() // retry batch-requests on interval
-		case result := <-completed:
-			if result.err != nil {
-				// batch failed, RPC may be broken, abort
-				return fmt.Errorf("batch request failed: %w", result.err)
+		case reqElem, ok := <-ibc.scheduled:
+			if !ok { // no more requests to do
+				return io.EOF
 			}
-			// if any element failed, add it to the requests for re-attempt
-			requests = append(requests, result.failed...)
-			awaited -= result.success
+			batch = append(batch, reqElem)
+			continue
 		case <-ctx.Done():
+			for _, r := range batch {
+				ibc.scheduled <- r
+			}
 			return ctx.Err()
+		default:
+			break
+		}
+		break
+	}
+
+	if err := ibc.getBatch(ctx, batch); err != nil {
+		for _, r := range batch {
+			ibc.scheduled <- r
+		}
+		return fmt.Errorf("failed batch-retrieval: %w", err)
+	}
+	var result error
+	for _, elem := range batch {
+		if elem.Error != nil {
+			result = multierror.Append(result, elem.Error)
+			elem.Error = nil // reset, we'll try this element again
+			ibc.scheduled <- elem
+			continue
+		} else {
+			atomic.AddUint32(&ibc.completed, 1)
+			if atomic.LoadUint32(&ibc.completed) >= uint32(len(ibc.requestsKeys)) {
+				close(ibc.scheduled)
+				return io.EOF
+			}
 		}
 	}
+	return result
+}
+
+// Complete indicates if the batch call is done.
+func (ibc *IterativeBatchCall[K, V, O]) Complete() bool {
+	ibc.resetLock.RLock()
+	defer ibc.resetLock.RUnlock()
+	return atomic.LoadUint32(&ibc.completed) >= uint32(len(ibc.requestsKeys))
+}
+
+// Result returns the fetched values, checked and transformed to the final output type, if available.
+// If the check fails, the IterativeBatchCall will Reset itself, to be ready for a re-attempt in fetching new data.
+func (ibc *IterativeBatchCall[K, V, O]) Result() (O, error) {
+	ibc.resetLock.RLock()
```

Comment:
Any reason to not defer runlock here?

Reply:
Because the later `Lock` during the later `Reset` and results-caching would conflict. Alternatively we can just wrap the function as a whole with `Lock` / `defer Unlock`, less optimal, but cleaner